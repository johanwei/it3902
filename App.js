import React, { useState, useMemo } from 'react';
import { StyleSheet, Text, View, Button } from 'react-native';
import MapView, { UrlTile, Polyline } from 'react-native-maps';
import * as FileSystem from 'expo-file-system';
import DownloadTiles from './components/DownloadTiles';
import TrackLocation from './components/TrackLocation';
import MarkLocations from './components/MarkLocations';
import RegisterSheep from './components/RegisterSheep';

const MAP_TYPE = Platform.OS == 'android' ? 'none' : 'standard'

export default function App() {

  const MAP_URL =  'https://opencache.statkart.no/gatekeeper/gk/gk.open_gmaps?layers=topo4&zoom={z}&x={x}&y={y}'
  const TILES_FOLDER = `${FileSystem.documentDirectory}tiles/{z}-{x}-{y}.png`
  const INITIAL_REGION = {
        latitude: 64.1608,
        longitude: 17.7808,
        latitudeDelta: 24.5742,
        longitudeDelta: 28.1047
      }

  const [isOffline, setIsOffline] = useState(false)
  const [visibleSettings, setVisibleSettings] = useState(false)
  const [mapRegion, setMapRegion] = useState(INITIAL_REGION)
  const [trackLocation, setTrackLocation] = useState(false)
  const [currentLocations, setCurrentLocations] = useState([])
  const [currentLocation, setCurrentLocation] = useState()
  const [stopTracking, setStopTracking] = useState(true)
  const [sheepLocation, setSheepLocation] = useState([{latitude: 69, longitude: 16}])


    
  const urlTemplate = useMemo(
    () =>
      isOffline
        ? `${TILES_FOLDER}`
        : `${MAP_URL}`,
    [isOffline]
  )

  async function deleteTiles() {
    try {
      await FileSystem.deleteAsync(`${FileSystem.documentDirectory}tiles`)
      alert('Slettet nedlastet kart')
    } catch (error) {
      console.warn(error)
    }
  }

  function toggleOffline() {
    setIsOffline(!isOffline)
  }

  function toggleDownloadSettings() {
    setVisibleSettings(!visisbleSettings)
  }

  function onDownloadComplete() {
    setIsOffline(true)
    setVisibleSettings(false)
  }

  function getCurrentLocations(locations){
    setCurrentLocations(locations);
  }

  function getCurrentLocation(loc){
    setCurrentLocation(loc);
  }

  function toggleTrackLocation() {
    setStopTracking(!stopTracking)
    setTrackLocation(!trackLocation);
  }
    
  return (
    
      <View style={styles.container}>
        <MapView 
        style={{ flex: 1}} 
        mapType={'standard'}
        maxZoomLevel={18}
        showsUserLocation={true}
        onRegionChangeComplete={setMapRegion}
        onLongPress={(e) => setSheepLocation([{latitude: e.nativeEvent.coordinate.latitude, 
                                               longitude: e.nativeEvent.coordinate.longitude}]) 
        /*setCurrentLocations([...currentLocations, {
          latitude: e.nativeEvent.coordinate.latitude,
          longitude: e.nativeEvent.coordinate.longitude,
          //setMarkers([...markers, {
          //latitude: e.nativeEvent.coordinate.latitude,
          //longitude: e.nativeEvent.coordinate.longitude,
          //id: markers.length,
          //value: e.nativeEvent.coordinate
        }])*/}
        >
        <UrlTile 
          urlTemplate={urlTemplate}
          shouldReplaceMapContent={false}
        />

        <MarkLocations locations = {currentLocations} />
           
        <RegisterSheep currentLocation = {currentLocation} sheepLocation = {sheepLocation} />
  
        {/*
        <MarkLocations />        
        <Polyline
            coordinates={currentLocations}
        strokeWidth={6}

        />
        {markers.map((marker) => {
          return (
        <MapView.Marker key={marker.id} coordinate={marker.value}>
          <MapView.Callout>
            <Text>ID: {marker.id}</Text>
            <Text>Latitude: {marker.value.latitude}</Text>
            <Text>Longitude: {marker.value.longitude}</Text>
          </MapView.Callout>        
          </MapView.Marker>)})}*/}

      </MapView>
      {visibleSettings && (
      <DownloadTiles mapRegion={mapRegion} onFinish={onDownloadComplete} />)}

      <View style={styles.actionContainer}>
        <Button raised title={'Last ned kart'} onPress={toggleDownloadSettings} />
        <Button raised title={'Slett kart'} onPress={deleteTiles} />
        <Button raised title={isOffline ? 'Bruk online' : 'Bruk offline'} onPress={toggleOffline}/>
      </View>      


      <Button raised title={stopTracking ? 'Start ny oppsynstur' : 'Avslutt oppsynstur'} onPress={toggleTrackLocation}/>

      
      {<TrackLocation listOfLocations={getCurrentLocations} stopTracking={stopTracking} currentLocation={getCurrentLocation} />}
    </View>
  )   
}

const styles = StyleSheet.create({
  actionContainer: {
    flexDirection: 'row',
    padding: 15,
    paddingTop: 30,
    zIndex: 999,
    justifyContent: 'space-around',
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
  },
  button: {
    backgroundColor: 'white',
    borderRadius: 10,
  },
  container: {
    flex: 1
  }
})