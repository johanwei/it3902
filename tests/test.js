import React from 'react';
import renderer from 'react-test-renderer';
import App from '../components/SheepNote.js';


it("test", () => {
    expect(1).toEqual(1)
})

 
test('Renders snapshot as expected', () => {
        const tree = renderer.create(<App />).toJSON();
        expect(tree).toMatchSnapshot();
    });