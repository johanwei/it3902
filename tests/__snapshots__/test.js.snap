// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders snapshot as expected 1`] = `
<View>
  <Modal
    animationType="none"
    hardwareAccelerated={false}
    onHide={[Function]}
    renderToHardwareTextureAndroid={true}
    transparent={true}
    visible={true}
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "black",
          "bottom": 0,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
    <View
      needsOffscreenAlphaCompositing={true}
      pointerEvents="box-none"
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
          "opacity": 0,
          "transform": Array [
            Object {
              "scale": 1.2,
            },
          ],
        }
      }
    >
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "marginTop": 22,
            },
            Object {
              "paddingBottom": 0,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "borderRadius": 13,
                "flexDirection": "column",
                "overflow": "hidden",
                "width": 270,
              },
              Object {},
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "backgroundColor": Object {
                    "semantic": Array [
                      "systemGray6",
                    ],
                  },
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
                Object {},
              ]
            }
          />
          <View
            style={
              Array [
                Object {
                  "margin": 18,
                },
                Object {},
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": Object {
                      "semantic": Array [
                        "label",
                      ],
                    },
                    "fontSize": 18,
                    "fontWeight": "600",
                    "textAlign": "center",
                  },
                  undefined,
                ]
              }
            >
              Registrer sau
            </Text>
          </View>
          <View
            style={
              Array [
                Object {
                  "backgroundColor": Object {
                    "semantic": Array [
                      "systemGray5",
                    ],
                  },
                  "borderColor": Object {
                    "semantic": Array [
                      "separator",
                    ],
                  },
                  "borderRadius": 6,
                  "borderWidth": 0.5,
                  "marginBottom": 20,
                  "marginHorizontal": 20,
                  "paddingHorizontal": 8,
                },
                undefined,
              ]
            }
          >
            <TextInput
              allowFontScaling={true}
              multiline={true}
              numberOfLines={4}
              onChangeText={[Function]}
              placeholder="Skriv notat her"
              placeholderTextColor={
                Object {
                  "semantic": Array [
                    "placeholderText",
                  ],
                }
              }
              rejectResponderTermination={true}
              style={
                Array [
                  Object {
                    "color": Object {
                      "semantic": Array [
                        "label",
                      ],
                    },
                  },
                  undefined,
                  Object {
                    "height": 74,
                  },
                ]
              }
              underlineColorAndroid={
                Object {
                  "semantic": Array [
                    "@android:color/hint_foreground_light",
                  ],
                }
              }
            />
          </View>
          <View
            style={
              Array [
                Object {
                  "borderTopColor": Object {
                    "semantic": Array [
                      "separator",
                    ],
                  },
                  "borderTopWidth": 0.5,
                  "flexDirection": "row",
                  "height": 46,
                  "justifyContent": "space-between",
                },
                Object {},
              ]
            }
          >
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "justifyContent": "center",
                  "opacity": 1,
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "backgroundColor": "transparent",
                      "color": Object {
                        "semantic": Array [
                          "link",
                        ],
                      },
                      "fontSize": 17,
                      "textAlign": "center",
                    },
                    Object {
                      "color": "#007ff9",
                      "fontWeight": "normal",
                    },
                    undefined,
                  ]
                }
              >
                Avbryt
              </Text>
            </View>
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": Object {
                      "semantic": Array [
                        "separator",
                      ],
                    },
                    "height": "100%",
                    "width": 0.5,
                  },
                  Object {},
                ]
              }
            />
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "justifyContent": "center",
                  "opacity": 1,
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "backgroundColor": "transparent",
                      "color": Object {
                        "semantic": Array [
                          "link",
                        ],
                      },
                      "fontSize": 17,
                      "textAlign": "center",
                    },
                    Object {
                      "color": "#007ff9",
                      "fontWeight": "normal",
                    },
                    undefined,
                  ]
                }
              >
                Registrer
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </Modal>
</View>
`;
